compiler/rustc_passes/src/dead.rs

```diff
@@ -780,14 +910,30 @@ impl<'tcx> Visitor<'tcx> for DeadVisitor<'tcx> {
     // This visitor should only visit a single module at a time.
     fn visit_mod(&mut self, _: &'tcx hir::Mod<'tcx>, _: Span, _: hir::HirId) {}
 
+    fn visit_enum_def(
+        &mut self,
+        enum_definition: &'tcx hir::EnumDef<'tcx>,
+        generics: &'tcx hir::Generics<'tcx>,
+        item_id: hir::HirId,
+        _: Span,
+    ) {
+        intravisit::walk_enum_def(self, enum_definition, generics, item_id);
+        self.warn_dead_fields_and_variants(item_id, "constructed", false)
+    }
+
     fn visit_variant(
         &mut self,
         variant: &'tcx hir::Variant<'tcx>,
         g: &'tcx hir::Generics<'tcx>,
         id: hir::HirId,
     ) {
         if self.should_warn_about_variant(&variant) {
-            self.warn_dead_code(variant.id, variant.span, variant.ident.name, "constructed");
+            self.store_dead_field_or_variant(
+                variant.id,
+                variant.span,
+                variant.ident.name,
+                self.tcx.lint_level_at_node(lint::builtin::DEAD_CODE, id).0,
+            );
         } else {
             intravisit::walk_variant(self, variant, g, id);
         }
```

Comment:
It would mean modifying `visit_variant` here so it doesn't call `store_dead_field_or_variant`, and modify `visit_enum_def` to look at the enum's variants itself and call `store_dead_field_or_variant` there while passing in the enum def. Same thing for `visit_field_def` (which would likely require more changes, as more things have fields).

Maybe instead modify the visitor itself to hold "current item being visited", instead of passing it as an arg in `store_dead_field_or_variant`?

Reply:
Is it like this?

```rust
fn visit_enum_def(
        &mut self,
        enum_definition: &'tcx hir::EnumDef<'tcx>,
        generics: &'tcx hir::Generics<'tcx>,
        item_id: hir::HirId,
        _: Span,
) {
    intravisit::walk_enum_def(self, enum_definition, generics, item_id);
    let dead_variants = enum_definition.variants
        .iter().filter(|variant| self.should_warn_about_variant(&variant)
        .map(|variant| DeadVariant {
            hir_id: variant.id,
            span: variant.span,
            name: variant.ident.name,
            level: self.tcx.lint_level_at_node(lint::builtin::DEAD_CODE, id).0,
            is_field: false,
        });
    self.warn_dead_fields_and_variants(item_id, "constructed", dead_variants)
}
```

Reply:
I refactored.